atomate/vasp/database.py

```diff
@@ -243,47 +208,112 @@ def insert_gridfs(self, d, collection="fs", compress=True, oid=None, task_id=Non
+        """
+        Insert the given document into a Maggma store, first check if the store is already
+
+        Args:
+            data: the document to be stored
+            collection (string): the name prefix for the maggma store
+            compress (bool): Whether to compress the data or not
+            oid (ObjectId()): the _id of the file; if specified, it must not already exist in GridFS
+            task_id(int or str): the task_id to store into the gridfs metadata
+        Returns:
+            file id, the type of compression used.
+        """
+        oid = oid or str(ObjectId())
+        compression_type = None
+
+        # make sure the store is availible
+        if collection not in self.maggma_stores:
+            self.get_obj_store(collection)
+
+        doc = {
+            "fs_id": oid,
+            "maggma_store_type": self.maggma_stores[collection].__class__.__name__,
+            "compression": compression_type,
+            "data": data
+        }
+
+        search_keys = ["fs_id", ]
+        if task_id is not None:
+            doc['task_id'] = task_id
+        elif isinstance(data, dict) and "task_id" in data:
+            search_keys.append('task_id')
+            doc['task_id'] = data['task_id']
+
+        # make sure the store is availible
+        with self.maggma_stores[collection] as store:
+            ping_ = store.index._collection.database.command("ping")
+            assert ping_.get("ok", 0) == 1.0, f"Not connected to the index store of {self.__name__}.maggma_store[{collection}]"
+            if isinstance(store, S3Store):
+                # TODO find some way to ping the aws service
+                # ping_ = self.maggma_stores[collection].s3_bucket._name
+                pass
+
+            if store.compress:
+                compression_type = "zlib"
+                doc['compression'] = 'zlib'
+
+            # print(doc, self.maggma_stores[collection].key)
+            store.update([doc], search_keys)
+
+        return oid, compression_type
+
+    def get_data_from_maggma_or_gridfs(self, task_id, key):
+        """
+        look for a task, then the object of type key associated with that task
+        Returns:
+            The data stored on object storage, typically a dictionary
+        """
         m_task = self.collection.find_one({"task_id": task_id}, {"calcs_reversed": 1})
-        fs_id = m_task["calcs_reversed"][0]["bandstructure_fs_id"]
-        fs = gridfs.GridFS(self.db, "bandstructure_fs")
-        bs_json = zlib.decompress(fs.get(fs_id).read())
-        bs_dict = json.loads(bs_json.decode())
-        if bs_dict["@class"] == "BandStructure":
-            return BandStructure.from_dict(bs_dict)
-        elif bs_dict["@class"] == "BandStructureSymmLine":
-            return BandStructureSymmLine.from_dict(bs_dict)
+        fs_id = m_task["calcs_reversed"][0][f"{key}_fs_id"]
+        obj_dict = None
+        if self._maggma_store_type in VALID_STORES:
```

Comment:
This whole logic can be written. Just ask for the store using `self.get_store` and if it is None go straight to gridfs.